prefect_dbt/cloud/jobs.py

```diff
@@ -318,9 +325,167 @@ def my_flow():
             f"Triggered job run with ID {run_id} was cancelled."
         )
     elif final_run_status == DbtCloudJobRunStatus.FAILED:
-        raise DbtCloudJobRunFailed(f"Triggered job run with ID: {run_id} failed.")
+        while retry_filtered_models_attempts > 0:
+            logger.info(
+                f"Retrying job run with ID: {run_id} "
+                f"{retry_filtered_models_attempts} more times"
+            )
+            try:
+                retry_filtered_models_attempts -= 1
+                run_data = await (
+                    retry_dbt_cloud_job_run_subset_and_wait_for_completion(
+                        dbt_cloud_credentials=dbt_cloud_credentials,
+                        run_id=run_id,
+                        trigger_job_run_options=trigger_job_run_options,
+                        max_wait_seconds=max_wait_seconds,
+                        poll_frequency_seconds=poll_frequency_seconds,
+                        retry_status_filters=retry_status_filters,
+                        retry_downstream_nodes=retry_downstream_nodes,
+                    )
+                )
+                return run_data
+            except Exception:
+                pass
+        else:
+            raise DbtCloudJobRunFailed(f"Triggered job run with ID: {run_id} failed.")
     else:
         raise RuntimeError(
             f"Triggered job run with ID: {run_id} ended with unexpected"
             f"status {final_run_status.value}."
         )
+
+
+def _filter_model_names_by_status(
+    run_artifact: Dict[str, Any], retry_status_filters: List[str]
+):
+    """
+    Filters model names from model results by status.
+    """
+    model_results = run_artifact.get("results", [])
+
+    model_names = set()
+    for model_result in model_results:
+        if model_result.get("status") in retry_status_filters:
+            model_id = model_result["unique_id"]
+            model_name = model_id.split(".")[-1]
+            model_names.add(model_name)
+
+    if len(model_names) == 0:
+        raise ValueError(
+            f"No valid model names were found using the filters: {retry_status_filters}"
+        )
+    return model_names
+
+
+def _build_trigger_job_run_options(
+    retry_downstream_nodes: bool,
+    model_names: List[str],
+    trigger_job_run_options: Optional[TriggerJobRunOptions] = None,
+):
+    """ """
+    graph_operator = "+ " if retry_downstream_nodes else " "
+    retry_command = f"dbt build --select {graph_operator.join(model_names)}"
```

Comment:
Is the intention that this retry functionality only be used for jobs that are constructed with `dbt build`?  I'm wondering what the behavior would be if a user had a job constructed with `['dbt run', 'dbt test']` and the failure occurs at `dbt test`?

Another thing to consider here is how to handle scenarios where users are passing variables in their dbt build command - `dbt build --vars '{env: prod}'`.  Those `vars` would be lost in this example and the command that's built would be run without valuable context.